### PR TITLE
Enforce Rust edition 2021

### LINKED ISSUE #2709: Upgrade Zebra to Rust 2021 edition
## Motivation

Rust will have a new 2021 edition in October:
https://doc.rust-lang.org/edition-guide/rust-2021/index.html

Zebra can stay on the 2018 edition, but we might miss out on new features. And some code will be harder to write.

## Tasks

- run the fix command for 2021 edition changes
- configure all crates to use the 2021 edition
    - this also upgrades the cargo dependency resolver to version 2
- ensure compatibility by adding `#[deny(clippy::rust-2021-compatibility)]` to our clippy config in each crate

### PR DESCRIPTION
**Pull request by dconnolly**
## Motivation

We've wanted to comply with Rust 2021 edition for a bit: 
https://github.com/ZcashFoundation/zebra/issues/2709

## Solution

Upgraded the `Cargo.toml`s to `edition = 2021`, did a little `cargo fix --edition` in each crate, ran tests, did a little `clippy fix`, and only had to tweak a few lines. 🎉

Resolves https://github.com/ZcashFoundation/zebra/issues/2709
## Review

Anyone can review, maybe @jvff @conradoplg 

### Reviewer Checklist

  - [ ] All tests and clippy pass

### CODE REVIEW

**Review by jvff — APPROVED** [review]

**Comment by gustavovalverde** [comment]
Mmm...this must have been captured by Mergify, as all the conditions are met. If possible, please do not merge it manually so I can check what's going on.

**Comment by jvff** [comment]
> Mmm...this must have been captured by Mergify, as all the conditions are met. If possible, please do not merge it manually so I can check what's going on.

Is it because there are two reviewers in the PR, and only one approval?

**Review by conradoplg — APPROVED** [review]
Looks good, I'll approve to check if it makes any difference to mergify...

**Comment by gustavovalverde** [comment]
@Mergifyio refresh

**Pull request merged**
